cscs-checks/prgenv/cuda/cuda_samples.py

```diff
@@ -0,0 +1,165 @@
+import reframe as rfm
+import reframe.utility.sanity as sn
+
+
+class CudaSamples(rfm.RegressionTest):
+    def __init__(self):
+        super().__init__()
+        self.valid_systems = ['daint:gpu', 'dom:gpu', 'tiger:gpu',
+                              'arolla:cn', 'tsa:cn',
+                              'ault:amdv100', 'ault:intelv100']
+        if self.current_system.name in ['arolla', 'tsa']:
+            self.valid_prog_environs += ['PrgEnv-pgi',
+                                         'PrgEnv-gnu-nompi',
+                                         'PrgEnv-pgi-nompi']
+            self.modules = ['cuda/10.1.243']
+        elif self.current_system.name in ['ault']:
+            self.valid_prog_environs = ['PrgEnv-gnu']
+            self.modules = ['cuda/11.0']
+        else:
+            self.valid_prog_environs = ['PrgEnv-cray',
+                                        'PrgEnv-gnu',
+                                        'PrgEnv-pgi']
+            self.modules = ['craype-accel-nvidia60']
+
+        if self.current_system.name in ['arolla', 'tsa', 'ault']:
+            self.exclusive_access = True
+            self.nvidia_sm = '70'
+        else:
+            self.nvidia_sm = '60'
+            self.modules = ['cudatoolkit']
+
+        self.sourcesdir = None
+        self.build_system = 'Make'
+        if self.current_system.name in ['daint']:
+            self.prebuild_cmds = ['export CUDA_HOME=$CUDATOOLKIT_HOME']
+        else:
+            self.prebuild_cmds = []
+
+        self.build_system.options = [
+            'SMS="%s"' %
+            self.nvidia_sm,
+            'CUDA_PATH=$CUDA_HOME']
+        self.maintainers = ['JO']
+        self.tags = {'production', 'external_resosurces'}
+
+
+@rfm.required_version('>=2.14')
+@rfm.simple_test
+class CudaDeviceQueryCheck(CudaSamples):
+    def __init__(self):
+        super().__init__()
+        self.descr = 'CUDA deviceQuery test.'
+        self.sourcesdir = 'https://github.com/NVIDIA/cuda-samples.git'
+        self.num_tasks = 1
+        self.prebuild_cmds += ['git checkout v11.0',
+                               'cd Samples/deviceQuery']
+        self.executable = 'Samples/deviceQuery/deviceQuery'
+        self.sanity_patterns = sn.assert_found(
+            r'Result = PASS',
+            self.stdout)
+
+
+class DependentCudaSamples(CudaSamples):
+    def __init__(self):
+        super().__init__()
+        self.depends_on('CudaDeviceQueryCheck')
+
+
+@rfm.required_version('>=2.14')
+@rfm.simple_test
+class CudaConcurrentKernelsCheck(DependentCudaSamples):
+    def __init__(self):
+        super().__init__()
+        self.descr = 'CUDA concurrentKernels test'
+        self.sanity_patterns = sn.assert_found(
+            r'Test passed',
+            self.stdout)
+
+    @rfm.require_deps
+    def set_prebuild_cmds(self, CudaDeviceQueryCheck):
+        self.prebuild_cmds += ['cd %s' % os.path.join(
+            CudaDeviceQueryCheck().stagedir,
+            'Samples', 'concurrentKernels')]
```

Comment:
It's strange that you have to cd in the other test's stage directory? Couldn't you just use it's executable?

Reply:
The entire repo is cloned in the CudaDeviceQueryCheck directory, so the "cd" is necessary to compile the code there. Alternatively, one could just copy the source to the directory of the current case. Does that sound better?

Reply:
I see, but the problem is that the test is no more self contained. It works with the async execution policy just because you clone the repo multiple times, one per each programming environment. If #1481 is merged, you would be able to have a single test running on a login partition that fetches the repo and the rest of the tests to depend on that. In that case, though, your current approach wouldn't work because the tests would start to step on each other during the compilation phase. So for the moment I would rather clone the repo in each test (the effect is practically the same to what you have now I think), but the tests will be self-contained.

Reply:
OK, got it. I just changed all these checks to clone the repo on each test. As expected, this makes the whole check quite a bit slower, so perhaps it's worth revisiting this in the future once #1481 is merged, and we can then try and come up with a way to reduce all these calls to `git clone` in here.
On the coding style side of things, I know that all these checks could just be parametrised, but I think that would take a bit of a dent in terms of readability.

Reply:
I hope to merge #1481 by next week or the week after, so we can then review the structure. Regarding parametrised vs. classes, I don't have a strong feeling. Perhaps it's better as you have it now.